src/raw/unix.rs

```diff
@@ -337,7 +359,54 @@ impl Socket {
         }
     }
 
-    ///Accepts incoming connection.
+    ///Accept a new incoming client connection and return its files descriptor and address.
+    ///
+    ///By default the newly created socket will be inheritable by child processes and created
+    ///in blocking I/O mode. This behaviour can be customized using the `flags` parameter:
+    ///
+    /// * `AcceptFlags::NON_BLOCKING`    – Mark the newly created socket as non-blocking
+    /// * `AcceptFlags::NON_INHERITABLE` – Mark the newly created socket as not inheritable by client processes
+    ///
+    ///Depending on the operating system's availablility of the `accept4(2)` system call this call
+    ///either pass the flags on to the operating system or emulate the call using `accept(2)`.
+    pub fn accept4(&self, flags: c_int) -> io::Result<(Socket, net::SocketAddr)> {
+        #[cfg(any(target_os = "linux", target_os = "freebsd", target_os = "openbsd", target_os = "dragonflybsd"))]
+        unsafe {
+            let mut storage: sockaddr_storage = mem::zeroed();
+            let mut len = mem::size_of_val(&storage) as socklen_t;
+
+            match accept4(self.inner, &mut storage as *mut _ as *mut _, &mut len, flags) {
+                SOCKET_ERROR => Err(io::Error::last_os_error()),
+                sock @ _ => {
+                    let addr = sockaddr_to_addr(&storage, len)?;
+                    Ok((Socket { inner: sock, }, addr))
+                }
+            }
+        }
+        
+        #[cfg(not(any(target_os = "linux", target_os = "freebsd", target_os = "openbsd", target_os = "dragonflybsd")))]
+        {
+            if (flags & !(libc::SOCK_NONBLOCK | libc::SOCK_CLOEXEC)) != 0 {
+                // `accept4()` also returns `EINVAL` if it receives an unknown flag
+                return Err(io::Error::from_raw_os_error(libc::EINVAL));
+            }
+
+            self.accept().map(|(sock, addr)| {
+                // Emulate the two most common (and useful) `accept4` flags using `ioctl`/`fcntl`
+                sock.set_nonblocking((flags & libc::SOCK_NONBLOCK) != 0).expect("Setting newly obtained client socket blocking mode");

```

Comment:
I'm not sure that panicking in case of impossibility to set  these flags is good idea.
While i do understand this case is highly unlikely i would prefer to guarantee that there is no way to panic in this lib.

Do you know what is behavior of accept4 in case of failing to set these flags?
Should we maybe return `io::Result<(Socket, net:;SocketAddr, io::Result<()>)>` and use third value of tuple to indicate result of settings these options?
That's the only graceful way i can think of.
There is no way we can accept and then close socket on faliure.